src/core/adapters/personalization.ts

```diff
@@ -0,0 +1,120 @@
+        }
+        break;
+      case Actions.SELECT_REFINEMENT:
+        return Selectors.refinementCrumb(state, payload.navigationId, payload.index);
+    }
+
+    return { field: undefined, value: undefined };
+  };
+
+  export const generateNewBias = () => ({ lastUsed: Math.floor(Date.now() / 1000) });
+
+  // tslint:disable-next-line max-line-length
+  export const transformToBrowser = (state: Store.Personalization.Biasing, reducerKey: string): BrowserStorageState => ({
+    allIds: state.allIds.map(({ variant, key }) => ({
+      variant,
+      key,
+      ...state.byId[variant][key]
+    }))
+  });
+
+  // tslint:disable-next-line max-line-length
+  export const transformFromBrowser = (incomingState: BrowserStorageState, state: Store.State): Store.Personalization.Biasing => {
+    const config = Selectors.config(state);
+    // tslint:disable-next-line max-line-length
+    const olderThanTime = Math.floor(Date.now() / 1000) - DAYS_IN_SECONDS * ConfigAdapter.extractRealTimeBiasingExpiry(config);
+    const filteredIncomingState = incomingState.allIds.filter((element) => element.lastUsed >= olderThanTime);
+    const biasingConfig = config.personalization.realTimeBiasing;
+    const allIds = [];
+    const byId = {};
+
+    filteredIncomingState.forEach(({ variant, key, lastUsed }) => {
+      if (!byId[variant]) {
+        byId[variant] = {}; // init
+      }
+      // do not push if too many biases
+      if (allIds.length < biasingConfig.maxBiases &&
+           // if no max defined for variant, push
+          (!(biasingConfig.attributes[variant] && biasingConfig.attributes[variant].maxBiases) ||
+           // if max defined and over, do not push
+           Object.keys(byId[variant]).length < biasingConfig.attributes[variant].maxBiases)) {
+        allIds.push({ variant, key });
+        byId[variant][key] = { lastUsed };
+      }
+    });
+
+    return { allIds, byId };
+  };
+
+  export const convertBiasToSearch = (state: Store.State) => {
+    const allIds = Selectors.realTimeBiasesAllIds(state);
+    const config = Selectors.config(state).personalization.realTimeBiasing;
+    const selectedRefinements = Selectors.selectedRefinements(state);
+
+    return allIds.filter(({ variant, key }) =>
+      config.attributes[variant] && !selectedRefinements.some(({ navigationName, type, value }) =>
+        type === 'Value' && navigationName === variant && key && value === key)
+    ).map(({ variant, key }) => ({
+      name: variant,
+      content: key,
+      strength: config.attributes[variant].strength || config.strength,
+    }));
+  };
+
+  // tslint:disable-next-line max-line-length
+  export const pruneBiases = (allIds: Store.Personalization.BiasKey[], variant: string, variantCount: number, config: Configuration.Personalization.RealTimeBiasing) => {
+    if (config.attributes[variant] && variantCount >= config.attributes[variant].maxBiases) {
+      for (let i = allIds.length - 1; i >= 0; i--) {
```

Comment:
this could be simplified to
```js
const index = allIds.reverse().findIndex((id) => id.variant === variant);
if (index !== -1) {
  return [...allIds.slice(0, index), ...allIds.slice(index + 1)];
}
```

Reply:
That mutates allIds though, which we can't do since we're using this on reducers. What we could do is declare a new variable like `const ids = allIds.slice();` at the start then mutate that, if you think that sounds alright.